### PR TITLE
MIR Body: Cache result of `is_cyclic` call

### PR DESCRIPTION
**Pull request by bugadani**

### CODE REVIEW

File: compiler/rustc_middle/src/mir/mod.rs [hunk 1]
```diff
@@ -353,7 +358,7 @@ impl<'tcx> Body<'tcx> {
     /// Returns `true` if a cycle exists in the control-flow graph that is reachable from the
     /// `START_BLOCK`.
     pub fn is_cfg_cyclic(&self) -> bool {
-        graph::is_cyclic(self)
+        self.is_cyclic.is_cyclic(self)
```
Comment (jonas-schievink) [inline compiler/rustc_middle/src/mir/mod.rs:361]: This needs to be invalidated whenever a terminator may change

Comment (bugadani) [reply compiler/rustc_middle/src/mir/mod.rs:361]: Thanks! Do you happen to have something in mind how I could cover that with a test?

Comment (jonas-schievink) [reply compiler/rustc_middle/src/mir/mod.rs:361]: Not really, outside of maybe a unit test that just calls a bunch of functions and checks that the flag was invalidated?

**Comment by bugadani** [comment]
Well since I've basically copied the predecessor cache, I've also copied it's testing strategy - i.e., none. I'll try to think about a way to avoid having to eagerly invalidate the caches.

**Comment by JohnCSimon** [comment]
@bugadani  - ping from triage - can you post your status on this PR? I've assigned it S-waiting-on-author because it looks like your PR still needs more work.

**Comment by bugadani** [comment]
@JohnCSimon I considered this PR "done", We could do a perf run to see if this change has any practical benefits, but otherwise I don't intend to make more modifications.

> looks like your PR still needs more work.

Do you have anything particular in mind?

**Comment by crlf0710** [comment]
Switching back to S-waiting-on-review.

**Comment by jyn514** [comment]
Up to -.8% instruction counts on many-assoc-items. No regressions :)

**Comment by oli-obk** [comment]
@bors r+

Thanks! Once we get a third cache we should start thinking about a more generic concept for caches, but at this stage, copying seems absolutely right to me.

**Pull request merged**